models/DataObject/ClassDefinition/Data/ReverseManyToManyObjectRelation.php

```diff
@@ -296,6 +278,249 @@ public function isOptimizedAdminLoading(): bool
     {
         return true;
     }
+
+    /**
+     * @param DataObject\Concrete $object
+     * @param array $params
+     *
+     * @return null
+     */
+    public function load($object, $params = [])
+    {
+        $db = Db::get();
+        $data = null;
+
+        if (!method_exists($this, 'getLazyLoading') or !$this->getLazyLoading() or (array_key_exists('force', $params) && $params['force'])) {
+            $relations = $db->fetchAll('SELECT * FROM object_relations_' . $this->getOwnerClassId() . " WHERE dest_id = ? AND fieldname = ? AND ownertype = 'object'", [$object->getId(), $this->getOwnerFieldName()]);
+        } else {
+            return null;
+        }
+
+        $relations = array_map(function ($relation) {
+            $relation['dest_id'] = $relation['src_id'];
+            unset($relation['src_id']);
+            return $relation;
+        }, $relations);
+
+        // using PHP sorting to order the relations, because "ORDER BY index ASC" in the queries above will cause a
+        // filesort in MySQL which is extremely slow especially when there are millions of relations in the database
+        usort($relations, static function ($a, $b) {
+            if ($a['index'] == $b['index']) {
+                return 0;
+            }
+
+            return ($a['index'] < $b['index']) ? -1 : 1;
+        });
+
+        $data = $this->loadData($relations, $object, $params);
+        if ($object instanceof DataObject\DirtyIndicatorInterface) {
+            $object->markFieldDirty($this->getName(), false);
+        }
+
+        return $data['data'];
+    }
+
+    /**
+     * @param DataObject\Concrete $object
+     * @param                     $data
+     * @param array               $params
+     *
+     * @return array|null
+     * @throws \InvalidArgumentException
+     */
+    public function preSetData($object, $data, $params = [])
+    {
+        $oldRelations = (array)$object->get($this->getName());
```

Comment:
general comment about preSetData: what is your opinion regarding the performance topic ?

Reply:
In my tests it is now much faster than with the old implementation in DataObjectController because many database updates for the owner objects are made with bulk UPDATE queries - in contrast to calling the save() method for every owner object.
Nevertheless I did not dare to remove the hint "The current object is not the owner of these relations, making changes here might make saving the object slow" ;-)